c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -284,6 +284,50 @@ impl<'tcx> TypeChecker<'tcx, '_> {
 
                     Some(Callee::MiscBuiltin) => {}
 
+                    Some(Callee::Malloc) => {
+                        self.visit_place(destination, Mutability::Mut);
+                    }
+
+                    Some(Callee::Calloc) => {
+                        self.visit_place(destination, Mutability::Mut);
+                    }
+                    Some(Callee::Realloc) => {
+                        self.visit_place(destination, Mutability::Mut);
+                        let pl_lty = self.acx.type_of(destination);
+                        assert!(args.len() == 2);
+                        self.visit_operand(&args[0]);
+                        let rv_lty = self.acx.type_of(&args[0]);
+
+                        // input needs FREE permission
+                        let perms = PermissionSet::FREE;
+                        self.constraints.add_all_perms(rv_lty.label, perms);
+
+                        // unify inner-most pointer types
+                        let mut rv_iter = rv_lty.iter();
+                        let mut pl_iter = pl_lty.iter();
+                        let _outermost_rv = rv_iter.next();
+                        let _outerpost_pl = pl_iter.next();
+                        // TODO: handle return types that are different:
+                        // realloc may return a *const void but take in *const *mut ...
```

Comment:
I don't think this comment applies.  C `realloc` takes and returns `*mut c_void` - if the C code passes a non-`void*` pointer, the Rust code will have explicit casts before and after the call.  And our custom `typed_realloc` for testing should similarly take and return pointers of the same type.

Reply:
removed in 1d0ae163